### ISSUE TITLE
Make functions/methods asynchronous

### ISSUE THREAD

**Issue by felipedau** [AnemoneLabs/unmessage#21]
During initial development (specially while maintaining two user interfaces), it was simpler to use predefined callbacks on unMessage's classes that were called after some result/action was ready. Although that currently works, `Deferreds` are a great solution for that and should be used. It will certainly make tests a lot easier.

**Comment by felipedau** [comment]
Hi @david415 and @meejah, to start working on this task I picked the path used when sending elements (the plaintext of packets: requests, texts, presence) to connect the callbacks with deferreds instead of predefined callback methods. You can find the changes in the [deferreds branch](https://github.com/felipedau/unmessage/tree/deferreds) and I suggest you looking at the commits sequentially (starting at 2037cc97f169d41e8f668e5af1f66e078e312bbb) to understand the steps I took to make these calls asynchronous.

I hope to use `@inlineCallbacks` wherever is possible, but when dealing with blocking code (probably at the very end of the chain) I will create a `Deferred`, start a thread that calls back when done and return the `Deferred` (e.g., [`UntalkSession.send_data`](https://github.com/felipedau/unmessage/blob/deferreds/unmessage/untalk.py#L226-L242)). Is it the best solution?

As you will see, conversations had two queues to send elements. One received element packets to be encrypted and another to send regular packets, after the encryption. I removed the former in the process, but I think that the latter may be useful when supporting multi-transport - I could be wrong and maybe `send_data` should be responsible by dealing with this though.

The "other half of the path" is done by the `ElementParser`. Previously, a callback was added to the second queue so that the thread sending the packet could notify when it was sent. This callback would notify the parser to check the type/content/sender of the element, act if needed, and finally call its respective predefined UI callback. It was useful because it dealt with both sent and received elements. Now that we have a proper callback chain, we do not need to parse the elements that are sent **in there**. In the current code, each method used to send an element is still adding a callback to the `parse` method, but the next step will be moving the sending logic from the parser to each of them, leaving only the receiving logic there.

Also, I thought that when raising errors in `@inlineCallbacks` decorated methods, they would be returned as failures. As it is not the case, I used `raise Failure(e)`. Is is correct?

Please, let me know if you think I am doing it right. Once I have your feedback, I will continue "moving the deferreds out" until they reach the UIs and I will know I can do the same for the other "not so asynchronous" paths that unMessage has.

Thanks!

**Comment by felipedau** [comment]
I went through the methods  of the `Peer` class and identified which should be returning a `Deferred`:

- [x] `Peer._send_presence`
- [x] `Peer._stop_tor`
- [x] `Peer._send_request`
- [x] `Peer._accept_request`
- [x] `Peer._untalk`
- [x] `Peer._send_message`
- [x] `Peer._authenticate`
- [x] `Peer.start`
    - [x] depends on `Peer._send_presence`
- [x] `Peer.stop`
    - [x] depends on `Peer._send_presence`
    - [x] depends on `Peer._stop_tor`
- [x] `Peer.send_request`
    - [x] depends on `Peer._send_request`
- [x] `Peer.accept_request`
    - [x] depends on `Peer._accept_request`
- [x] `Peer.untalk`
    - [x] depends on `Peer._untalk`
- [x] `Peer.send_message`
    - [x] depends on `Peer._send_message`
- [x] `Peer.authenticate`
    - [x] depends on `Peer._authenticate`

The methods above should also use `inlineCallbacks`. The following methods already return a `Deferred` but do not use `inlineCallbacks` yet:

- [x] `Peer._start_tor`
- [ ] ~~`Peer._connect`~~ (no need to use `inlineCallbacks`)
- [x] `Peer._start_server`

Thanks!

**Comment by meejah** [comment]
When dealing with APIs that you need to call from an async method, but those APIs are still threaded/blocking style, using Twisted's `deferToThread` method is probably the best approach. https://twistedmatrix.com/documents/current/api/twisted.internet.threads.deferToThread.html

**Comment by meejah** [comment]
You should just be able to e.g. `raise RuntimeError("something bad")` inside an `@inlineCallbacks` method and this will result in the `Deferred` that your method (now) returns getting `.errback()`'d with a `Failure`. (So it won't "return" a `Failure` but your errback chain will receive one).

**Comment by meejah** [comment]
Taking a quick pass through the branch, another subtle thing: the errback in a `.addCallbacks(cb, eb)` call will NOT receive any error from the "cb" itself. So you might want to do `d.addCallback(cb)` and then `d.addErrback(eb)` so that the errback will be notified from any errors in `cb`, too.

**Comment by felipedau** [comment]
> When dealing with APIs that you need to call from an async method, but those APIs are still threaded/blocking style, using Twisted's deferToThread method is probably the best approach. https://twistedmatrix.com/documents/current/api/twisted.internet.threads.deferToThread.html

Ahh, that's awesome! Can I do something like this then?

```python
@inlineCallbacks
def send_data(self, data):
    try:
        result = yield deferToThread(self.connection.send, (data,))
    except Exception as e:
        raise errors.UntalkError(
            message='{}: {}'.format(str(type(e)), e.message))
    else:
        returnValue(result)
```

In that specific case, I do not know what errors `send` can possibly raise, so converting any exception to an `UntalkError` (which subclasses `UnmessageError`) is a way to easily display any error to the UIs.

> You should just be able to e.g. raise RuntimeError("something bad") inside an @inlineCallbacks method and this will result in the Deferred that your method (now) returns getting .errback()'d with a Failure. (So it won't "return" a Failure but your errback chain will receive one).

Got it! I think that once I move the other half of the logic to each method, then simply using try/except to handle the exceptions is better. I opted to raise `Failure`s because `.check()` is so convenient :D But raising just exceptions is indeed simpler once you are in try/except blocks.

> Taking a quick pass through the branch, another subtle thing: the errback in a .addCallbacks(cb, eb) call will NOT receive any error from the "cb" itself. So you might want to do d.addCallback(cb) and then d.addErrback(eb) so that the errback will be notified from any errors in cb, too.

Right. I believe that the cases where I added `addCallbacks` are the following:

1. One `Deferred` is being chained into another (`d0.addCallbacks(d1.callback, d1.errback)`)
2. The current scope's method is going to change once I continue "moving the `Deferred`s out"

In the first case, I do not think that `d0.addCallback(d1.callback).addErrback(d1.errback)` would be correct because it wouldn't possibly know what the callback would do. Then leaving it up to whatever receives that `Deferred` seems better. Is that right?

In the second case, the callbacks are still notifying errors on their own. Until that is fixed, I'd prefer to see it crash than having an error being handled by an unrelated errback, if that makes sense.

As I am trying to use `@inlineCallbacks` everywhere is possible I think that `.addCallback(cb).addErrback(eb)` will be mostly present in the UIs or in the methods called by them.

What do you think?

Thanks!

**Comment by meejah** [comment]
yes, you can use defer-to-thread like that :)

**Comment by felipedau** [comment]
Hi @david415 and @meejah,

I am almost done with the [list](https://github.com/AnemoneLabs/unmessage/issues/21#issuecomment-305177317) of methods that should be
asynchronous. You can follow the changes in the PR #70.

You may notice I used this `join()` function:


```python
def join(d):
    results = [None, None]
    event = Event()

    def callback(result):
        results[0] = result
        event.set()

    def errback(failure):
        results[1] = failure
        event.set()

    d.addCallbacks(callback, errback)
    event.wait()

    if results[1]:
        results[1].raiseException()
    else:
        return results[0]
```

I know it's terrible, but it's useful for this transition. I hope to
remove it completely once this task is done.

I am currently working on passing a reactor to `Peer()` and finally
run an asynchronous test.

Let me know what you think!

**Comment by felipedau** [comment]
Hi @david415 and @meejah,

Here is an update:

- Finished making the methods in the [list](https://github.com/AnemoneLabs/unmessage/issues/21#issuecomment-305177317) asynchronous
- Installed the reactor in the GUI (89702f9a0157cf8021194286d70a19d694970fa1)
- Installed the reactor in the CLI (44d0ee8af6fbce71f0c7ba048799aec4c6fe908e)
- Removed support for optionally passing the reactor to `Peer()` and launching/calling it from a thread as now both UIs pass one (449011f5a1837d9621603b4faddb6165bd7faf1d)

To test that passing a reactor works, I used this to start a peer (bootstrap Tor and create an Onion Service):

```python
@react
@inlineCallbacks
def main(reactor):
    peer = Peer('pytest', reactor)
    notification = yield peer.start(begin_log=True,
                                    begin_log_std=True,
                                    log_level=LogLevel.debug)
    assert str(notification) == 'Peer started'
    yield peer.stop()
```

As well as ran an actual test using pytest-twisted:


```python
@pytest.inlineCallbacks
def test_start(reactor):
    peer = Peer('pytest', reactor)
    notification = yield peer.start()
    assert str(notification) == 'Peer started'
    yield peer.stop()

```

However, I do not think we want to do this kind of tests right now. Unit testing first.

I believe that #70 can be merged and for the next days I will continue working on the tests, making some changes (some parts should not depend on IO so that it is possible to just make objects talk to other objects) and actually writing tests.

As we split the asynchronous and test tasks, for the next period of GSoC I plan to work on the receiving parts of unMessage that still use threads (and that `join()` function) and also continue testing.

Let me know what you think!

**Comment by felipedau** [comment]
Now that #70 and #77 are merged, this task is complete.
